src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,289 @@
+<?php
+
+namespace React\Socket;
+
+use React\Dns\Model\Message;
+use React\Dns\Resolver\ResolverInterface;
+use React\EventLoop\LoopInterface;
+use React\EventLoop\TimerInterface;
+use React\Promise;
+use React\Promise\CancellablePromiseInterface;
+
+/**
+ * @internal
+ */
+final class HappyEyeBallsConnectionBuilder
+{
+    const CHECK_INTERVAL = 0.1;
+
+    public $loop;
+    public $connector;
+    public $resolver;
+    public $uri;
+    public $host;
+    public $resolved = array(
+        Message::TYPE_A    => false,
+        Message::TYPE_AAAA => false,
+    );
+    public $resolverPromises = array();
+    public $connectionPromises = array();
+    public $connectQueue = array();
+    public $timer;
+    public $parts;
+    public $ipsCount = 0;
+    public $failureCount = 0;
+    public $resolve;
+    public $reject;
+
+    public function __construct(LoopInterface $loop, ConnectorInterface $connector, ResolverInterface $resolver, $uri, $host, $parts)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+        $this->uri = $uri;
+        $this->host = $host;
+        $this->parts = $parts;
+    }
+
+    public function connect()
+    {
+        $that = $this;
+        return new Promise\Promise(function ($resolve, $reject) use ($that) {
+            $lookupResolve = function ($type) use ($that, $resolve, $reject) {
+                return function (array $ips) use ($that, $type, $resolve, $reject) {
+                    unset($that->resolverPromises[$type]);
+                    $that->resolved[$type] = true;
+
+                    $that->mixIpsIntoConnectQueue($ips);
+
+                    $that->check($resolve, $reject);
+                };
+            };
+
+            $that->resolverPromises[Message::TYPE_AAAA] = $that->resolve(Message::TYPE_AAAA, $resolve, $reject)->then($lookupResolve(Message::TYPE_AAAA));
+            $that->resolverPromises[Message::TYPE_A] = $that->resolve(Message::TYPE_A, $resolve, $reject)->then(function ($ips) use ($that) {
+                if ($that->resolved[Message::TYPE_AAAA] === true) {
+                    return Promise\resolve($ips);
+                }
+
+                /**
+                 * Delay A lookup by 50ms sending out connection to IPv4 addresses when IPv6 records haven't
+                 * resolved yet as per RFC.
+                 *
+                 * @link https://tools.ietf.org/html/rfc8305#section-3
+                 */
+                return Promise\Timer\resolve(0.05, $that->loop)->then(function () use ($ips) {
+                    return Promise\resolve($ips);
+                });
```

Comment:
Updated implementation looks much better to me!

The way I understand this piece of code is to now *always* wait for 50ms after the `A` DNS resolution before starting the first IPv4 connection. It's my understanding this should a maximum wait to at which point an IPv4 connection will be tried even when the `AAAA` query is still pending. Accordingly, if the `AAAA` query returns no results after 1ms, it should start trying IPv4 connections right after this response.

Reply:
Updated it to immediately after `AAAA` has been resolved release `A` for it's 50ms wait.